### PR TITLE
feat(semantic): translate FlatBuffers semantic graph to Go

### PR DESCRIPTION
**Pull request by wolffcm**
This PR commits code that translates semantic graph FlatBuffers to the Go "natural" semantic graph.

I wrote a tool called `fbgen` that does generates the code to do the deserialization.  It uses reflection to find all the implementors of `Node` in the Go `semantic` package, and generates a `FromBuf` method for them.  The code generation is performed using the `jen` package, same as what we do for the Flux standard library.

The generated code is in `flatbuffers_gen.go`.  Helper functions that were easier to write by hand are in `flatbuffers.go`.  

There is not much testing here yet, which is why coverage is so poor.  When serialization of semantic graph in Rust is complete, then I can follow up with more tests that show that the semantic graph produced by Go matches the semantic graph produced by Rust and deserialized from the FlatBuffers representation.

### CODE REVIEW

File: internal/cmd/fbgen/cmd/semantic.go [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+	flagOutput            string
+	flagContinueOnMissing bool
+)
+
+func init() {
+	rootCmd.AddCommand(generateCmd)
+	generateCmd.Flags().StringVar(&flagOutput, "output", "",
+		"When present, output will be written to the specified file. Otherwise, output is piped to stdout.")
+	generateCmd.Flags().BoolVar(&flagContinueOnMissing, "continue-on-missing", false,
+		"When set, TODO comments will be put in generated code for missing types, instead of terminating with an error.")
```
[the remark this answers is not in the document]
Comment (fchikwekwe) [reply internal/cmd/fbgen/cmd/semantic.go]: @yzhang1991 isn't this a TODO? I thought this was going to be filled in from other places in the code. Unless I'm misunderstanding.

File: internal/cmd/fbgen/cmd/semantic.go [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+		)
+		for i := 0; i < st.NumFields(); i++ {
+			c, err := doField(o, st.Field(i))
+			if err != nil {
+				return err
+			}
+			if c != nil {
+				cs = append(cs, c...)
+			}
+		}
```
Comment (fchikwekwe) [inline internal/cmd/fbgen/cmd/semantic.go]: I'm not totally understanding this section, but I'm guessing that this is going through the files and looking for places to generate code based on the function definitions below.

File: internal/cmd/fbgen/cmd/semantic.go [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+	"Extern":                     {},
+	"ExternBlock":                {},
+	"ExternalVariableAssignment": {},
+	"InterpolatedPart":           {},
+	"TextPart":                   {},
+	"FunctionExpression":         {},
+	"FunctionBlock":              {},
+	"FunctionParameter":          {},
+	"FunctionParameters":         {},
+}
```
Comment (fchikwekwe) [inline internal/cmd/fbgen/cmd/semantic.go]: Are we storing the skipped nodes here?

File: semantic/flatbuffers.go [hunk 4]
```diff
@@ -0,0 +1,559 @@
...
+			return nil, err
+		}
+		return s, nil
+	case fbsemantic.StatementReturnStatement:
+		fbStmt := new(fbsemantic.ReturnStatement)
+		fbStmt.Init(tbl.Bytes, tbl.Pos)
+		s := &ReturnStatement{}
+		if err := s.FromBuf(fbStmt); err != nil {
+			return nil, err
+		}
```
Comment (fchikwekwe) [inline semantic/flatbuffers.go]: This is pretty cool. I wish there were some solutions like this for Rust.

File: semantic/flatbuffers_test.go [hunk 5]
```diff
@@ -1,57 +1,244 @@
...
 	flatbuffers "github.com/google/flatbuffers/go"
+	"github.com/google/go-cmp/cmp"
+	"github.com/influxdata/flux/parser"
 	"github.com/influxdata/flux/semantic/internal/fbsemantic"
 )
 
-func TestFlatBuffers(t *testing.T) {
+// TODO(cwolff): There needs to be more testing here.  Once we can serialize an arbitrary semantic graph
+//   in Rust, we can get more complete coverage without having to create FlatBuffers by hand.
+
```
Comment (fchikwekwe) [inline semantic/flatbuffers_test.go]: Hopefully I'll have the Rust stuff done in a few days so we can see some more testing here.

File: internal/cmd/fbgen/cmd/semantic.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (wolffcm) [reply internal/cmd/fbgen/cmd/semantic.go]: I put this in to help me as work was in progress.  If a semantic graph node is not handled by the tool (that's what I meant by "missing type"), then a TODO comment is put into the generated code.  That way I could easily find all the nodes that I had not yet done.

I'll update the comment to be clearer.

File: Makefile [hunk 6]
```diff
@@ -52,7 +54,7 @@ ast/internal/fbast: ast/ast.fbs
 libflux/src/ast/flatbuffers/ast_generated.rs: ast/ast.fbs
 	flatc --rust -o libflux/src/ast/flatbuffers ast/ast.fbs
 
-semantic/internal/fbsemantic: semantic/semantic.fbs
+semantic/internal/fbsemantic semantic/flatbuffers_gen.go: semantic/semantic.fbs semantic/graph.go internal/cmd/fbgen/cmd/semantic.go
```
[the remark this answers is not in the document]
Comment (wolffcm) [reply Makefile]: `make` will just look at the timestamps of the files on the right of the `:` and rebuild using the given command if any of them are newer than the targets on the left.  In this case I want to force a rebuild when:
- The flatbuffers schema changes
- Any of the Go semantic graph types change
- The `fbgen` tool that generates the deserialization code changes

**Comment by wolffcm** [comment]
I updated the branch with lots more comments, hopefully it makes everything clearer.

**Comment by wolffcm** [comment]
@yzhang1991 I am on the fence about it.  I am leaning towards merging this PR, but leaving issue #2091 open, and only closing once we have roundtrip testing.  Does that sound reasonable?

**Comment by yzhang1991** [comment]
@wolffcm Yeah, I agree. Please update the issue description so people know there is another PR coming for the roundtrip test.

**Pull request merged**